### PR TITLE
feat: Add --full-description option to include full comment in schema

### PR DESCRIPTION
**Pull request by alexchexes**
This adds support for a new CLI/config option `--raw-jsdoc` (`rawJsDoc`), which includes a `rawJsDoc` field in the generated JSON Schema. This field preserves the full JSDoc comment block *as is* (after removing asterisks), without any formatting, tag filtering, inheritance, etc.

- Adds CLI and config option: `--raw-jsdoc` / `rawJsDoc`
- Adds `rawJsDoc` keyword to Ajv validator
- Enhances `ExtendedAnnotationsReader` to extract raw JSDoc when available
- Includes thorough tests in `test/config/jsdoc-raw`
- Updates documentation (README)

Related to discussion in #1544 and complements `--markdown-description` (#1773)

#### Typescript:
```ts
/**
 * **Summary**
 * Link {@link https://example.com}
 * @description Short summary
 * @see Reference
 * @see https://example.com
 * - Additional comment
 */
```
#### Output:
```json
{
  "description": "Short summary",
  "markdownDescription": "**Summary**\nLink  {@link  https://example.com }",
  "rawJsDoc": "**Summary**\nLink {@link https://example.com}\n@description Short summary\n@see Reference\n@see https://example.com\n- Additional comment"
}
```

# Version

Published prerelease version: `v2.5.0-next.2`

<details>
  <summary>Changelog</summary>

  :tada: This release contains work from new contributors! :tada:
  
  Thanks for all your work!
  
  :heart: Alex ([@alexchexes](https://github.com/alexchexes))
  
  :heart: Valentyne Stigloher ([@pixunil](https://github.com/pixunil))
  
  #### 🚀 Enhancement
  
  - feat: Add --full-description option to include full comment in schema [#2224](https://github.com/vega/ts-json-schema-generator/pull/2224) ([@alexchexes](https://github.com/alexchexes))
  - feat(parser): support SpreadElement in array literals [#2269](https://github.com/vega/ts-json-schema-generator/pull/2269) ([@alexchexes](https://github.com/alexchexes))
  
  #### 🐛 Bug Fix
  
  - fix: correctly generate anyOf on unions with string and boolean constant [#2208](https://github.com/vega/ts-json-schema-generator/pull/2208) ([@pixunil](https://github.com/pixunil))
  - fix: fully unwrap union aliases in mapped keys to avoid generating incorrect additionalProperties [#2232](https://github.com/vega/ts-json-schema-generator/pull/2232) ([@alexchexes](https://github.com/alexchexes))
  - fix: avoid incorrect additionalProperties for Pick<..., AliasLiteralUnion> [#2230](https://github.com/vega/ts-json-schema-generator/pull/2230) ([@alexchexes](https://github.com/alexchexes))
  
  #### 🔩 Dependency Updates
  
  - chore(deps-dev): bump typescript-eslint from 8.32.1 to 8.33.1 [#2270](https://github.com/vega/ts-json-schema-generator/pull/2270) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.15.29 to 22.15.30 [#2271](https://github.com/vega/ts-json-schema-generator/pull/2271) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.27.0 to 9.28.0 [#2264](https://github.com/vega/ts-json-schema-generator/pull/2264) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/core from 7.27.1 to 7.27.4 [#2265](https://github.com/vega/ts-json-schema-generator/pull/2265) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @eslint/js from 9.27.0 to 9.28.0 [#2266](https://github.com/vega/ts-json-schema-generator/pull/2266) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint-plugin-prettier from 5.4.0 to 5.4.1 [#2267](https://github.com/vega/ts-json-schema-generator/pull/2267) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.15.21 to 22.15.29 [#2268](https://github.com/vega/ts-json-schema-generator/pull/2268) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.15.18 to 22.15.21 [#2262](https://github.com/vega/ts-json-schema-generator/pull/2262) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-env from 7.27.1 to 7.27.2 [#2263](https://github.com/vega/ts-json-schema-generator/pull/2263) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps): bump commander from 13.1.0 to 14.0.0 [#2255](https://github.com/vega/ts-json-schema-generator/pull/2255) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump typescript-eslint from 8.32.0 to 8.32.1 [#2254](https://github.com/vega/ts-json-schema-generator/pull/2254) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.26.0 to 9.27.0 [#2256](https://github.com/vega/ts-json-schema-generator/pull/2256) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @eslint/js from 9.26.0 to 9.27.0 [#2257](https://github.com/vega/ts-json-schema-generator/pull/2257) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.15.3 to 22.15.18 [#2258](https://github.com/vega/ts-json-schema-generator/pull/2258) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint-config-prettier from 10.1.2 to 10.1.5 [#2249](https://github.com/vega/ts-json-schema-generator/pull/2249) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump typescript-eslint from 8.31.0 to 8.32.0 [#2250](https://github.com/vega/ts-json-schema-generator/pull/2250) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump tsx from 4.19.3 to 4.19.4 [#2251](https://github.com/vega/ts-json-schema-generator/pull/2251) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.25.1 to 9.26.0 [#2252](https://github.com/vega/ts-json-schema-generator/pull/2252) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint-plugin-prettier from 5.2.6 to 5.4.0 [#2253](https://github.com/vega/ts-json-schema-generator/pull/2253) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-typescript from 7.27.0 to 7.27.1 [#2244](https://github.com/vega/ts-json-schema-generator/pull/2244) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @eslint/js from 9.25.1 to 9.26.0 [#2245](https://github.com/vega/ts-json-schema-generator/pull/2245) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-env from 7.26.9 to 7.27.1 [#2246](https://github.com/vega/ts-json-schema-generator/pull/2246) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/core from 7.26.10 to 7.27.1 [#2247](https://github.com/vega/ts-json-schema-generator/pull/2247) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.15.2 to 22.15.3 [#2248](https://github.com/vega/ts-json-schema-generator/pull/2248) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.14.1 to 22.15.2 [#2239](https://github.com/vega/ts-json-schema-generator/pull/2239) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps): bump glob from 11.0.1 to 11.0.2 [#2240](https://github.com/vega/ts-json-schema-generator/pull/2240) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.25.0 to 9.25.1 [#2242](https://github.com/vega/ts-json-schema-generator/pull/2242) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump typescript-eslint from 8.30.1 to 8.31.0 [#2243](https://github.com/vega/ts-json-schema-generator/pull/2243) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump typescript-eslint from 8.29.1 to 8.30.1 [#2235](https://github.com/vega/ts-json-schema-generator/pull/2235) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @eslint/js from 9.24.0 to 9.25.0 [#2236](https://github.com/vega/ts-json-schema-generator/pull/2236) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.24.0 to 9.25.0 [#2237](https://github.com/vega/ts-json-schema-generator/pull/2237) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 22.14.0 to 22.14.1 [#2225](https://github.com/vega/ts-json-schema-generator/pull/2225) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump typescript-eslint from 8.29.0 to 8.29.1 [#2226](https://github.com/vega/ts-json-schema-generator/pull/2226) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint-config-prettier from 10.1.1 to 10.1.2 [#2227](https://github.com/vega/ts-json-schema-generator/pull/2227) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 9.23.0 to 9.24.0 [#2228](https://github.com/vega/ts-json-schema-generator/pull/2228) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  
  #### Authors: 3
  
  - [@dependabot[bot]](https://github.com/dependabot[bot])
  - Alex ([@alexchexes](https://github.com/alexchexes))
  - Valentyne Stigloher ([@pixunil](https://github.com/pixunil))
</details>

### CODE REVIEW

**Comment by arthurfiorette** [comment]
I already did not like going out of the JSDoc spec with `markdownDescription`. Not sure how I feel about another description-related field. What about `rawJsDoc` being a flag that puts the raw JSdoc contents inside the description, so we don't create a new field?

File: src/Utils/getRawJsDoc.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            return i;
+        } else if (!isWhitespaceCharCode(charCode)) {
+            break;
+        }
+    }
+    return -1;
+}
+
+function getTextWithoutStars(inputText: string) {
+    const innerTextWithStars = inputText.replace(/^\/\*\*[^\S\n]*\n?/, "").replace(/(\r?\n)?[^\S\n]*\*\/$/, "");
```
Comment (arthurfiorette) [inline src/Utils/getRawJsDoc.ts]: Could you do it without regex?

File: src/Utils/getRawJsDoc.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+function isWhitespaceCharCode(charCode: number) {
+    return regExWhitespaceSet.has(charCode);
+}
+
+function getStarPosIfFirstNonWhitespaceChar(text: string) {
+    for (let i = 0; i < text.length; i++) {
+        const charCode = text.charCodeAt(i);
+
+        if (charCode === CharCodes.ASTERISK) {
```
Comment (arthurfiorette) [inline src/Utils/getRawJsDoc.ts]: why charcode compare when you can just `text[i] == '*'`?

File: src/Utils/getRawJsDoc.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            return i;
+        } else if (!isWhitespaceCharCode(charCode)) {
```
Comment (arthurfiorette) [inline src/Utils/getRawJsDoc.ts]: you can `text.trim()` to avoid having to deal with custom whitespace char and reduce the amount of chars to loop.

**Review by arthurfiorette — APPROVED** [review]

**Review by arthurfiorette — CHANGES_REQUESTED** [review]
Wrongfully approved.

**Review by domoritz — COMMENTED** [review]
Instead of another command, we should combine it with --markdown-description and make the option an enum.

**Comment by arthurfiorette** [comment]
Awesome, enum value seems better :)

**Comment by alexchexes** [comment]
> What about rawJsDoc being a flag that puts the raw JSdoc contents inside the description, so we don't create a new field?

That was actually my first idea too — to reuse the `description` field for raw JSDoc — but while exploring it, I realized it forces users to choose between either the parsed `description` (with tags stripped and formatting normalized) or the raw JSDoc as originally written.

Replacing `description` would break the default behavior and potentially affect tooling or users who rely on it for readability or validation, yet those users (just like myself) could also need the full, unformatted, and unstripped JSDoc.

So, while I initially wasn't a big fan of introducing a new field either, adding a separate `rawJsDoc` field turned out to be the safest and least disruptive option.

@domoritz  
Regarding the enum idea — do you mean something like this?
```
--jsdoc-description parsed      # only includes description (default)
--jsdoc-description markdown    # includes description + markdownDescription
--jsdoc-description raw         # includes description + rawJsDoc
--jsdoc-description full        # includes all: description + markdownDescription + rawJsDoc
```
That would definitely simplify the config surface and avoid introducing another CLI flag, so if we all agree this is a good approach, I can rework the PR in that direction.

File: src/Utils/getRawJsDoc.ts (hunk 2 above)
Comment (alexchexes) [reply src/Utils/getRawJsDoc.ts]: Yep, I agree, I'll change that. I guess originally (it's from ts-morph) it used `charCodeAt` as a "micro-optimization" since it's in a loop.

File: src/Utils/getRawJsDoc.ts (hunk 3 above)
Comment (alexchexes) [reply src/Utils/getRawJsDoc.ts]: I'll experiment with using `trim()` (or `trimStart()`) to simplify the loop, but I'll need to write and run tests first to verify that we preserve whitespace in all kinds of multiline comments exactly as before (I was relying on the ts-morph impl since it is battle-tested)

File: src/Utils/getRawJsDoc.ts (hunk 1 above)
Comment (alexchexes) [reply src/Utils/getRawJsDoc.ts]: I'm open to replacing the regex with another approach, but could you clarify why regex is considered a problem here, since any alternative solution would be significantly more verbose? Handling single-line and multi-line comments differently would require adding many extra conditions.

**Comment by arthurfiorette** [comment]
It's an opt-in feature, so it shouldn't break anything by default. The user should know why they chose it. 

`jsdoc-description` seems a good name fit. 

---

It would be good if we could remove `--markdown-description` from the documentation but still interpret it, just as if it were `--jsdoc-description markdown` alongside it, emit [a deprecation warning](https://nodejs.org/api/process.html#processemitwarningwarning-options).

File: src/Utils/getRawJsDoc.ts (hunk 3 above)
Comment (alexchexes) [reply src/Utils/getRawJsDoc.ts]: done

File: src/Utils/getRawJsDoc.ts (hunk 1 above)
Comment (alexchexes) [reply src/Utils/getRawJsDoc.ts]: Please take a look at the latest version and getRawJsDoc unit-test cases (to understand what this regex is covering) to confirm whether the question is still relevant

**Comment by alexchexes** [comment]
> It's an opt-in feature, so it shouldn't break anything by default. The user should know why they chose it.

By "break the default behavior," I meant that overwriting the `description` with the raw JSDoc would change what tools and users might expect (even when they opt in for raw JSDoc) — a short, clean summary string without special at-tags, etc.

There are real cases where both are useful: for example, showing a concise `description` inline, while exposing the full `rawJsDoc` in a tooltip, collapsible developer panel, or markdown preview.

If we allow only one at a time, then in such cases the user would have to generate the schema twice with different options and invent a strategy to merge them. That doesn't feel like a clean approach. At the same time, I can imagine cases where you **do** want the raw JSDoc to go into the `description` field — for example, if a tool only supports `description` and you can't make it recognize custom fields like `rawJsDoc`. Definitely a bit of a dilemma...

> `jsdoc-description` seems like a good name fit.

Agreed! The only question is: what option values should we support?  
Something like `parsed | markdown | raw | full` as described above?

**Comment by alexchexes** [comment]
> Agreed! The only question is: what option values should we support?
> Something like `parsed | markdown | raw | full` as described above?

So, what are we doing? 🙂

**Comment by domoritz** [comment]
```
--jsdoc-description parsed      # only includes description (default)
--jsdoc-description markdown    # includes description + markdownDescription
--jsdoc-description raw         # includes description + rawJsDoc
--jsdoc-description full        # includes all: description + markdownDescription + rawJsDoc
```

Hmm, after seeing this, I think actually what you have (no enum) makes more sense. It lets users enable and disable whatever they need.

File: README.md [hunk 4]
```diff
@@ -47,6 +47,7 @@ By default, the command-line generator will use the `tsconfig.json` file in the
   -e, --expose <expose>          Type exposing (choices: "all", "none", "export", default: "export")
   -j, --jsDoc <extended>         Read JsDoc annotations (choices: "none", "basic", "extended", default: "extended")
   --markdown-description         Generate `markdownDescription` in addition to `description`.
+  --raw-jsdoc                    Include the full raw JSDoc comment as `rawJsDoc` in the schema.
```
Comment (domoritz) [inline README.md]: Should this be `--raw-jsDoc` since we have `--jsDoc` already?

**Comment by domoritz** [comment]
> It would be good if we could remove --markdown-description from the documentation but still interpret it, just as if it were --jsdoc-description markdown alongside it, emit [a deprecation warning](https://nodejs.org/api/process.html#processemitwarningwarning-options).

Or do we want to get rid of `markdown-description` in favor of this option? Then it might be worth revisiting all the options we have related to descriptions and make sure they are understandable. I like having fewer config options.

File: README.md (hunk 4 above)
Comment (alexchexes) [reply README.md]: Honestly, I would replace `--jsDoc` with `--jsdoc` or `--js-doc`, but that's up to you guys

**Comment by arthurfiorette** [comment]
Some users might rely on markdown description, but that's probably because we don't yet provide raw JSdoc. As soon as we provide raw JSdoc, it's up to the user to parse the raw JSDoc into markdown however they want...

**Comment by alexchexes** [comment]
> Some users might rely on markdown description, but that's probably because we don't yet provide raw JSdoc. As soon as we provide raw JSdoc, it's up to the user to parse the raw JSDoc into markdown however they want...

So, is the plan to [deprecate](https://nodejs.org/api/process.html#processemitwarningwarning-options) `--markdown-description` for now and recommend using the `rawJsDoc` option instead?  
Are we adding a `rawJsDoc` field that will eventually replace the `markdownDescription` field?

**Comment by arthurfiorette** [comment]
Up to @domoritz decide :)

File: README.md (hunk 4 above)
Comment (domoritz) [reply README.md]: I'd too if we started over but let's not break it now.

**Comment by alexchexes** [comment]
Hi @domoritz! Just checking in - let us know your thoughts so we can wrap this up!

**Comment by domoritz** [comment]
Thanks for the ping. Just got back from some long travel. Let's get this wrapped up. I'm in favor of eventually deprecating the markdown parsing. Right now, it's still handy as it removes the jsdoc annotations which people don't need in markdown descriptions intended for editors.

Let's do this

* [ ] rename the option to `--full-description`. Describe in the comment that it's for the full, raw description from the comment in ts. Also happy with `--raw-description` if you prefer. 
* [ ] rename the internal variable to `fullDescription`. Also happy with `rawDescription`.
* [ ] make sure the command line argument pipes through

**Comment by alexchexes** [comment]
Hi! Hope you had a nice trip! 

> Let's do this ...

Done!

**Comment by domoritz** [comment]
Thank you!
